### PR TITLE
Smoother and safer template project from init command.

### LINKED ISSUE #237: Built charm defaults to kubernetes cloud
When you run `charmcraft init` it builds a charm with a default series being `kubernetes` but a default beginner Juju bootstrap is usually not.

A simple fix might be to expand the TODO items to include modifying the series array in the `metadata.yaml`. To hopefully avoid this error:

`ERROR cannot add application "my-charm": series "kubernetes" in a non container model not valid`

In the linked issue it suggests a few options for commands to build charms specifically for k8s but maybe the init command should be interactive instead.

Related: https://github.com/canonical/charmcraft/issues/111

### PR DESCRIPTION
**Pull request by facundobatista**
This involves several details:

- Do NOT build the charm if the project has leftover TODOs from when it was bootstrapped with `init` .
- For that, started to use `TEMPLATE-TODO` to make it very distinct.
- Added a couple of those to different places where the developer also needs to pay attention.
- Left the templates in an state that will create a "solid" charm project.

Fixes #237, #238, #243 and #247.

### CODE REVIEW

**Review by cjp256 — COMMENTED** [review]
What about just warning on TODOs rather than trying to hard error?

It does seem a little odd that the example is uncommented now in the template.  Was there an issue with the comment-only file?

**Comment by jnsgruk** [comment]
I agree that warning about TEMPLATE-TODO items would be nicer than failing the build (thinking about first run, Hello, World style experiences with charmcraft)

**Comment by facundobatista** [comment]
@cjp256 @jnsgruk: we want a hard error because the developers could totally ignore the warnings and just build and push a very low quality charm to Charmhub. We want to lower the barrier as much as possible for people creating charms, but at the same time avoid as much as possible to saturate Charmhub with those just-bootstrapped charms. See issue #197 from Mark, for example.

Regarding the un-commented example: that is how it born, we commented it out at some point in time (see same issue I mentioned above) so people doesn't publish a charm with real (but bogus) actions and config without realizing it. In this new model, after `init` bootstrapping, you have a real charm, that has that functionality, and even passing tests for those examples, but you HAVE to pay attention to all that (at least to consciously remove the TODO tokens or the whole examples).

**Comment by jnsgruk** [comment]
Is it possible to restrict this at the point of upload/publish, rather than build?

**Comment by facundobatista** [comment]
> Is it possible to restrict this at the point of upload/publish, rather than build? It seems for the purposes of getting people started with development, you should be able to quite simply follow something like:
> 
> ```shell
> mkdir my-charm && cd my-charm
> charmcraft init
> charmcraft buiild
> juju deploy ./my-charm.charm
> ```

This is a great idea. I moved the detection from build to upload. I also made it so it flags *all* files that needs to be fixed (not the first one), as the loop of changing the file and detecting the "problem" is bigger.

File: charmcraft/templates/init/src/charm.py.j2 [hunk 1]
```diff
@@ -33,20 +33,30 @@ class {{ class_name }}(CharmBase):
         self._stored.set_default(things=[])
 
     def _on_install(self, _):
-        # Perform any install steps and report status
+        """Perform any install steps and report status."""
         self.unit.status = ActiveStatus("Active and running")
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply charmcraft/templates/init/src/charm.py.j2:37]: I like it showing that a message can be given, as it's the only example in the file showing how to set the unit status.

[the remark this answers is not in the document]
Comment (jnsgruk) [reply charmcraft/templates/init/src/charm.py.j2:37]: Maybe less of a problem now there are actually some developer docs? :rofl:

[the remark this answers is not in the document]
Comment (jnsgruk) [reply charmcraft/templates/init/src/charm.py.j2:37]: We could remove the message for now, but put a comment that links to the docs?

**Comment by cjp256** [comment]
I'd also suggest reconsidering how "init" works.  If it were more interactive, charmcraft could help the user create a more tailored template as needed.

**Review by cjp256 — APPROVED** [review]

File: charmcraft/commands/build.py [hunk 2]
```diff
@@ -87,6 +90,19 @@ def relativise(src, dst):
     return pathlib.Path(os.path.relpath(str(dst), str(src.parent)))
 
 
+def validate_template_is_handled(filepath):
+    """Check that the file has not a mark of coming from the template and wasn't handled.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply charmcraft/commands/build.py]: Improved the docstring in this function (moved to the upload phase).

File: charmcraft/commands/build.py [hunk 3]
```diff
@@ -87,6 +90,19 @@ def relativise(src, dst):
...
+
+    This is important to eliminate the risk of getting built charms that are just created
+    with the `init` commands and not paid attentiont.
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply charmcraft/commands/build.py]: Same, thanks!

**Comment by facundobatista** [comment]
> I'd also suggest reconsidering how "init" works. If it were more interactive, charmcraft could help the user create a more tailored template as needed.

Not a bad idea, for sure not for this PR.

File: charmcraft/commands/init.py [hunk 4]
```diff
@@ -105,12 +106,17 @@ def run(self, args):
         if not re.match(r"[a-z][a-z0-9-]*[a-z0-9]$", args.name):
             raise CommandError("{} is not a valid charm name".format(args.name))
 
+        if args.series is None:
+            series = "[kubernetes]  # TEMPLATE-TODO: fix to an Ubuntu series if not using k8s"
```
Comment (cmatsuoka) [inline charmcraft/commands/init.py:110]: Maybe "change to"?

File: charmcraft/commands/store/__init__.py [hunk 5]
```diff
@@ -322,9 +325,31 @@ def fill_parser(self, parser):
...
+        This is important to avoid uploading low-quality charms that are just
+        bootstrapped and not corrected.
+        """
+        # we're already sure we can open it ok
+        zf = zipfile.ZipFile(str(filepath))
+
+        tainted_filenames = []
+        for name in zf.namelist():
+            content = zf.read(name)
+            if INIT_TEMPLATE_TOKEN in content:
```
Comment (cmatsuoka) [inline charmcraft/commands/store/__init__.py:340]: Is it safe to assume that all files in the zipfile will be small enough to read to memory? Can we filter the possible candidates to have `TEMPLATE-TODO` by filename, or could it really be any file?

File: charmcraft/templates/init/config.yaml.j2 [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
-#     default: 🎁
-#     description: A thing used by the charm.
-#     type: string
+# TEMPLATE-TODO: change this example to suit your needs.
+# If you don't need a config, you can remove the file entirely.
+# It ties in to the example _on_config_changed handler in src/charm.py
+
+options:
+  thing:
+    default: 🎁
```
Comment (cmatsuoka) [inline charmcraft/templates/init/config.yaml.j2:10]: This is a nice default.

File: charmcraft/commands/init.py (hunk 4 above)
Comment (facundobatista) [reply charmcraft/commands/init.py:110]: I like it! improved.

File: charmcraft/commands/store/__init__.py (hunk 5 above)
Comment (facundobatista) [reply charmcraft/commands/store/__init__.py:340]: Charm files tend to be very small, yes. We *could* filter by files (e.g., *.py, *.yaml, *.md, etc), but I wanted to keep this as decoupled as possible from the other part of the code.

File: charmcraft/templates/init/config.yaml.j2 (hunk 6 above)
Comment (facundobatista) [reply charmcraft/templates/init/config.yaml.j2:10]: :)

**Comment by facundobatista** [comment]
functionality landed!

**Review by cmatsuoka — APPROVED** [review]
Thanks Facundo, it looks good. Maybe I'm not 100% comfortable with the possibility of reading large files to memory but this can be handled later if needed.

**Pull request merged**